### PR TITLE
Runtime: Handle synthesized decl "related entity" tags.

### PR DESCRIPTION
**Pull request by jckarter**
Clang-importer-synthesized declarations get an extra tag character included in their mangling, which was not being preserved in type context descriptors. This caused runtime lookup for these synthesized types to fail. Fix this by adding the tag information to type context descriptors and teaching the runtime to match it up when fetching metadata by mangled name. Fixes rdar://problem/40878715.

### CODE REVIEW

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by jckarter** [comment]
@swift-ci Please test

**Review by jrose-apple — APPROVED** [review]
Makes sense to me. Too bad about having to burn another tag bit on it.

**Comment by jckarter** [comment]
@swift-ci Please test

**Review by DougGregor — APPROVED** [review]
LGTM. It'd be nice for the RemoteAST path to get a test as well.

**Comment by jckarter** [comment]
@DougGregor Thanks! The existing RemoteAST tests exercise this behavior; now that we correctly include the related entity tag when we `_buildDemanglingForMetadata`, the `MetadataReader` now has to interpret it correctly.

**Comment by jckarter** [comment]
@swift-ci Please test

File: lib/RemoteAST/RemoteAST.cpp [hunk 1]
```diff
@@ -754,7 +754,13 @@ RemoteASTTypeBuilder::findForeignNominalTypeDecl(StringRef name,
       if (HadError) return;
       if (decl == Result) return;
       if (!Result) {
-        Result = cast<NominalTypeDecl>(decl);
+        if (auto resultAlias = dyn_cast<TypeAliasDecl>(decl)) {
+          Result = resultAlias->getUnderlyingTypeLoc().getType()
```
Comment (jrose-apple) [inline lib/RemoteAST/RemoteAST.cpp]: It might be best to limit this to when `resultAlias->isCompatibilityAlias()`.

**Comment by jckarter** [comment]
@swift-ci Please test

**Pull request merged**